File: fuentes/web/src-ts/camaras.ts


import { Assert, ComprErrorGL, Log } from "./utilidades.js"
import { Vec2, Vec3, Vec4, Mat4, CMat4 } from "./vec-mat.js"
import { Cauce } from "./cauce.js"

// -------------------------------------------------------------------------------------------------
// 

/**
 * Dimensiones del viewport (no hace falta!)
 */
export class Viewport 
{
    public ancho : number = 0 
    public alto  : number = 0 
    //private ratio_yx : number = 0.0  

    /**
     * Construye un objeto 'viewport' dando sus dimensiones
     * 
     * @param ancho_inicial (number) número de columnas de pixel del viewport (>0)
     * @param alto_inicial  (number) número de filas de pixel del viewport (>0)
     */
    constructor( ancho_inicial : number, alto_inicial : number )
    {
        this.fijarDimensiones( ancho_inicial, alto_inicial )
    }
    // -----------------------------------------------------------------------------
    
    /**
     * Cambia las dimensiones del viewport
     * @param nuevo_ancho (number) nuevo número de columnas de pixel del viewport (>0)
     * @param nuevo_alto  (number) nuevo número de filas de pixel del viewport (>0)
     */
    fijarDimensiones( nuevo_ancho : number, nuevo_alto : number ) : void
    {
        const nombref : string = 'Viewport.setDimensions()'
        
        Assert( 0 < nuevo_ancho && 0 < nuevo_alto, `${nombref} el ancho o alto del viewport no pueden ser negativos (las dimensiones son ${nuevo_ancho} x ${nuevo_alto})`)

        this.ancho    = nuevo_ancho
        this.alto     = nuevo_alto
    }
    // -----------------------------------------------------------------------------

    /**
     * Devuelve el aspect ratio Y/X de este viewport
     * @returns alto dividido por el ancho 
     */
    leerRatioYX() : number 
    {
        return this.alto / this.ancho 
    }
}
// ***************************************************************************************

export abstract class CamaraInteractiva
{
    /**
     * Nombre de la cámara (para depurar)
     */
    protected nombre : String = "(no asignado)"

    /**
     * Viewport de esta cámara (se usa para saber sus proporciones) 
     */
    protected viewport : Viewport = new Viewport( 256, 256 )

    /**
     * Matriz de vista 
     */
    protected mat_vista : Mat4 = CMat4.ident()

    /**
     * Matriz de proyección 
     */
    protected mat_proyeccion : Mat4 = CMat4.ident()

    //
    /**
     * Construye una cámara nueva, por defecto es una cámara perspectiva
     * @param nombre 
     */
    constructor( nombre? : string )
    {
        if ( nombre != undefined )
            this.nombre = nombre
    }

    /**
     * Modifica los ángulos horizonal y vertical de la cámara orbital 
     * (debe ser redefinido en clases derivadas) 
     * @param delta_x (number) -- incremento en horizontal 
     * @param delta_y (number) -- incremento en vertical 
     */
    abstract mover( delta_x : number, delta_y : number ) : void ;

    /**
     * hacer 'zoom', usando un factor de escala relativo [factor_rel]
     * (debe ser redefinido en clases derivadas)
     */
    abstract zoom( factor_rel : number ) : void ;
    

    /**
     * Recalcular las matrices en función del estado actual
     * (debe ser redefinido en clases derivadas)
     */
    abstract recalcularMatrices( ) : void ;

    /**
     * fijar las dimensiones del viewport en el que se usará esta cámara
     * @param {Viewport} nuevo_viewport 
     */
    fijarViewport( nuevo_viewport : Viewport  ) : void
    {
        this.viewport = nuevo_viewport
    }

    /**
     * Activar esta cámara en un cauce.
     *   - Actualiza las matrices de vista y proyeccion, 
     *   - Fija los uniforms de matrices en el cauce.
     *   - Hace 'reset' (reinicializa) de la matriz de modelado en el cauce.
     * @param cauce (Cauce) cauce en el que se activa la matriz
     */
    activar( cauce : Cauce )
    {
        this.recalcularMatrices()
        cauce.fijarMatrizVista( this.mat_vista )
        cauce.fijarMatrizProyeccion( this.mat_proyeccion )
        cauce.resetMM()
    }

}
// ******************************************************************************

/**
 * Clase que encapsula una matriz de proyección y una matriz de vista,
 * junto con las operaciones para rotarla con dos grados de libertad)
 */
export class CamaraOrbital3D extends CamaraInteractiva
{
    /**
     * amplitud de campo vertical (en grados) para cámaras perspectivas
     */
    private fovy_grad : number = 60.0

    /**
     * distancia al plano de recorte delantero 
     */
    private near : number = 0.05 

    /**
     * Distancia al plano de recorte trasero 
     */
    private far : number = 30.05 

     /**
      *  true si la cámara es perspectiva, false si es paralela (por ahora solo puede ser true)
      */
    private es_perspectiva : boolean = true  

    /**
     * Punto de atención actual
     */
    private punto_atencion : Vec3 = new Vec3([ 0.0, 0.0, 0.0 ])

    /**
     * Ángulo vertical de la cámara en grados (rotación entorno a X)
     */
    private angulo_vert_grad : number = 45.0 

    /**
     * Ángulo horizontal de la cámara en grados (rotación entorno a Y)
     */
    private angulo_hor_grad : number = 45.0

    /**
     * Distancia desde el observador hasta el punto de atención
     */
    private distancia : number = 3.0
    

    /**
     * Construye una cámara nueva, por defecto es una cámara perspectiva
     * @param nombre 
     */
    
    constructor( nombre? : string )
    {
       super( nombre ) 
    }

    /**
     * constructor por defecto, pone el nombre como "no asignado"
     */
   
    
    
    /**
     * Actualiza la matriz de proyección a partir del aspect ratio del viewport,
     * de la amplitud de campo y de los valores 'near' y 'far' actuales.
     */
    private recalcularMatrizProyeccion() : void 
    {
        const nombref : string = "Camara.actualizarMatrizProyeccion"
        Assert( this.es_perspectiva, `${nombref} todavía no gestiono cámaras paralelas, sorry...` )

        this.mat_proyeccion = CMat4.perspective( this.fovy_grad, 1.0/this.viewport.leerRatioYX(),
                                                    this.near, this.far )
    }

    /**
     * Actualiza la matriz de vista, a partir del punto de atención, los ángulos 
     * vertical y horizontal, y la distancia observador-punto de atención
     */
    private recalcularMatrizVista() : void 
    {
        const nombref : string = "Camara.actualizarMatrizProyeccion"
        const 
            a      : Vec3 = this.punto_atencion,
            mtras1 : Mat4 = CMat4.traslacion( new Vec3([ -a.x, -a.y, -a.z ])),
            mrot2  : Mat4 = CMat4.rotacionXgrad( this.angulo_vert_grad ),
            mrot1  : Mat4 = CMat4.rotacionYgrad( -this.angulo_hor_grad ),
            mtras2 : Mat4 = CMat4.traslacion( new Vec3([ 0, 0, -this.distancia ]))

        this.mat_vista = mtras2.componer( mrot2.componer( mrot1.componer( mtras1 ) ) )
        //Log(`${nombref} matriz de vista  == ${this.matriz_vista}`)        
    }

    /**
     * Actualiza las matrices de vista y proyección
     * (implementa el método abstracto de la clase padre)
     */
    recalcularMatrices(): void 
    {
        this.recalcularMatrizVista() 
        this.recalcularMatrizProyeccion()
    }

    /**
     * Modifica los ángulos horizonal y vertical de la cámara orbital 
     * 
     * @param delta_x (number) -- incremento en horizontal 
     * @param delta_y (number) -- incremento en vertical 
     */
    mover( delta_x : number, delta_y : number ) : void 
    {
        const nombref : string = 'CamaraOrbital.moverHV'

        this.angulo_hor_grad  += delta_x
        this.angulo_vert_grad += delta_y
    }

    /**
     * Modifica la distancia entre el punto de atención y el punto del observador
     * @param factor_rel (number) +1 o -1 según la dirección de movimiento del ratón
     */
    zoom( factor_rel : number ) : void
    {
        const nombref = 'CamaraOrbital.zoom'
        //Log(`${nombref} factor_rel == ${factor_rel}, dist = ${this.distancia}`)
        const incr_en_porcentaje : number = 3.0*factor_rel
        this.distancia = this.distancia * (1.0 + incr_en_porcentaje/100.0)
    }


}  // fin class CamaraOrbital3D

// ****************************************************************************************************

/**
 * Cámaras para vistas 2D, el plano de visión es el plano XY
 * En el viewport se visualiza un cuadrado con centro en [cv_centro_wcc] y
 * lado [cv_lado_wcc]  (todos los valores en coordenadas de mundo)
 */
export class CamaraVista2D extends CamaraInteractiva
{
    private cv_centro_wcc : Vec3   = new Vec3( [0.0, 0.0, 0.0] )
    private cv_lado_wcc   : number = 2.0

    /**
     * Construye una cámara nueva, por defecto es una cámara perspectiva
     * @param nombre 
     */
    constructor( nombre : string )
    {
        super( nombre ) 
    }

    // ---------------------------------------------------------------------------------------------
    /**
     * Mueve la cámara tras un evento de movimiento con un desplazamiento
     * proporcional a [delta_x] y [delta_y]
     * Desplaza [cv_centro_wcc] usando [delta_x] y [delta_y]
     */
    mover( delta_x : number, delta_y : number ) : void
    {
        this.cv_centro_wcc 
            = this.cv_centro_wcc.menos( new Vec3([ delta_x, delta_y, 0.0 ]) )
    }
    // ---------------------------------------------------------------------------------------------
    /**
     * hacer 'zoom', usando un factor de escala relativo [factor_rel]
     * cambia el lado del cuadrado visible ([cv_lado_wcc]
     */
    zoom( factor_rel : number ) : void 
    {
        this.cv_lado_wcc *= 1.0/factor_rel
    }
    // ---------------------------------------------------------------------------------------------

    recalcularMatrices()
    {
        let tam_pixel_wcc = this.cv_lado_wcc/ Math.min( this.viewport.ancho, this.viewport.alto ) // lado de un pixel en WCC
        let fx            = (2.0/this.cv_lado_wcc)* Math.min( 1.0, this.viewport.alto/this.viewport.ancho )
        let fy            = (2.0/this.cv_lado_wcc)* Math.min( 1.0, this.viewport.ancho/this.viewport.alto )

        this.mat_vista     = CMat4.traslacion( this.cv_centro_wcc.mult( - tam_pixel_wcc) )
        this.mat_proyeccion = CMat4.escalado( new Vec3([ fx, fy, 1.0 ]))

    }
}  // fin de CamaraVista2D

File: fuentes/moviles/proyecto.android-studio/app/src/main/java/mds/pcg1/camaras.kt
// *********************************************************************
// **
// ** Asignatura: PCG (Programación del Cauce Gráfico).
// **
// ** Clases relacionadas con cámaras interactivas
// ** Copyright (C) 2024 Carlos Ureña
// **
// ** Clase: CamaraInteractiva, CamaraVista2D, Camara3D
// **
// ** This program is free software: you can redistribute it and/or modify
// ** it under the terms of the GNU General Public License as published by
// ** the Free Software Foundation, either version 3 of the License, or
// ** (at your option) any later version.
// **
// ** This program is distributed in the hope that it will be useful,
// ** but WITHOUT ANY WARRANTY; without even the implied warranty of
// ** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// ** GNU General Public License for more details.
// **
// ** You should have received a copy of the GNU General Public License
// ** along with this program.  If not, see <http://www.gnu.org/licenses/>.
// **
// *********************************************************************

package mds.pcg1.camaras

import android.util.Log
import mds.pcg1.cauce.*
import mds.pcg1.utilidades.*
import mds.pcg1.vec_mat.*
import kotlin.math.min

// *************************************************************************************************

open class CamaraInteractiva( ancho_vp : Int, alto_vp : Int )
{
    // dimensiones del viewport sobre el cual se usa esta cámara
    protected var tam_vp_x_dcc  : Float = ancho_vp.toFloat()   // ancho del viewport en pixels
    protected var tam_vp_y_dcc  : Float = alto_vp.toFloat()    // alto del viewport en pixels

    // matrices de vista y proyección
    protected var mat_vista      = Mat4.ident()
    protected var mat_proyeccion = Mat4.ident()

    /**
     * Mueve la cámara tras un evento de movimiento con un desplazamiento
     * proporcional a [delta_x] y [delta_y]
     * (debe ser redefinido en clases derivadas)
     */
    open fun mover( delta_x : Float, delta_y : Float )
    {
        val TAGF = "[${object {}.javaClass.enclosingMethod?.name?: nfnd}]"
        Log.v( TAG, "este objeto cámara no tiene redefinido el método '$TAGF'")
    }

    /**
     * hacer 'zoom', usando un factor de escala relativo [factor_rel]
     * (debe ser redefinido en clases derivadas)
     */
    open fun zoom( factor_rel : Float )
    {
        val TAGF = "[${object {}.javaClass.enclosingMethod?.name?: nfnd}]"
        Log.v( TAG, "este objeto cámara no tiene redefinido el método '$TAGF'")
    }

    /**
     * Recalcular las matrices en función del estado actual
     * (debe ser redefinido en clases derivadas)
     */
    protected open fun recalcularMatrices( )
    {
        val TAGF = "[${object {}.javaClass.enclosingMethod?.name?: nfnd}]"
        Log.v( TAG, "este objeto cámara no tiene redefinido el método '$TAGF'")
    }

    /**
     * cambia las dimensiones del viewport (tiene efecto cuando se vuelve a activar)
     */
    fun fijarViewport( ancho_vp : Int, alto_vp : Int )
    {
        tam_vp_x_dcc  = ancho_vp.toFloat()   // ancho del viewport en pixels
        tam_vp_y_dcc  = alto_vp.toFloat()    // alto del viewport en pixels
    }

    /**
     * Activa esta cámara en este cauce: recalcula las matrices y fija uniforms
     */
    fun activar( cauce : CauceBase )
    {
        //val TAGF = "[${object {}.javaClass.enclosingMethod?.name?: nfnd}]"

        recalcularMatrices()
        cauce.fijarMatrizVista( mat_vista )
        cauce.fijarMatrizProyeccion( mat_proyeccion )
    }

}
// *************************************************************************************************

/**
 * Cámaras para vistas 2D, el plano de visión es el plano XY
 * En el viewport se visualiza un cuadrado con centro en [cv_centro_wcc] y
 * lado [cv_lado_wcc]  (todos los valores en coordenadas de mundo)
 */
class CamaraVista2D( ancho_vp : Int, alto_vp : Int ) : CamaraInteractiva( ancho_vp, alto_vp )
{
    private var cv_centro_wcc : Vec3  = Vec3( 0.0f, 0.0f, 0.0f )
    private var cv_lado_wcc   : Float = 2.0f

    // ---------------------------------------------------------------------------------------------
    /**
     * Mueve la cámara tras un evento de movimiento con un desplazamiento
     * proporcional a [delta_x] y [delta_y]
     * Desplaza [cv_centro_wcc] usando [delta_x] y [delta_y]
     */
    override fun mover( delta_x : Float, delta_y : Float )
    {
        //val TAGF = "[${object {}.javaClass.enclosingMethod?.name?: nfnd}]"

        cv_centro_wcc = cv_centro_wcc - Vec3( delta_x, delta_y, 0.0f )
    }
    // ---------------------------------------------------------------------------------------------
    /**
     * hacer 'zoom', usando un factor de escala relativo [factor_rel]
     * cambia el lado del cuadrado visible ([cv_lado_wcc]
     */
    override fun zoom( factor_rel : Float )
    {
        //val TAGF = "[${object {}.javaClass.enclosingMethod?.name?: nfnd}]"
        cv_lado_wcc *= 1.0f/factor_rel
    }
    // ---------------------------------------------------------------------------------------------

    override fun recalcularMatrices()
    {
        val tam_pixel_wcc = cv_lado_wcc/ min( tam_vp_x_dcc, tam_vp_y_dcc ) // lado de un pixel en WCC
        val fx            = (2.0f/cv_lado_wcc)* min( 1.0f, tam_vp_y_dcc/tam_vp_x_dcc )
        val fy            = (2.0f/cv_lado_wcc)* min( 1.0f, tam_vp_x_dcc/tam_vp_y_dcc )

        mat_vista = Mat4.traslacion( (-tam_pixel_wcc)*cv_centro_wcc )
        mat_proyeccion = Mat4.escalado( Vec3( fx, fy, 1.0f  ))

    }
}  // fin de CamaraVista2D

// *************************************************************************************************

class CamaraOrbital3D( ancho_vp : Int, alto_vp : Int ) : CamaraInteractiva( ancho_vp, alto_vp )
{

    // amplitud de campo vertical (en grados) para cámaras perspectivas
    private var fovy_grad : Float = 70.0f

    // distancia al plano de recorte delantero
    private var near : Float = 0.05f

    // distancia al plano de recorte trasero
    private var far : Float = 100.0f

    // true si la cámara es perspectiva, false si es paralela (por ahora solo puede ser true)
    //private es_perspectiva : boolean = true

    // punto de atención actual
    private var punto_atencion : Vec3 = Vec3( 0.0f, 0.0f, 0.0f )

    // ángulo vertical de la cámara en grados (rotación entorno a X)
    private var angulo_vert_grad : Float = 20.0f //45.0f

    // ángulo horizontal de la cámara en grados (rotación entorno a Y)
    private var angulo_hor_grad : Float = -20.0f //45.0f

    // distancia desde el observador hasta el punto de atención
    private var distancia : Float = 12.0f

    // ---------------------------------------------------------------------------------------------

    override fun recalcularMatrices()
    {
        val ratio_xy : Float = tam_vp_x_dcc/tam_vp_y_dcc
        val ratio_yx : Float = tam_vp_y_dcc/tam_vp_x_dcc
        mat_proyeccion = Mat4.perspective( fovy_grad, ratio_xy, near, far )

        val a      : Vec3 = punto_atencion
        val mtras1 : Mat4 = Mat4.traslacion( Vec3( -a.x, -a.y, -a.z ))
        val mrot2  : Mat4 = Mat4.rotacionXgrad( angulo_vert_grad )
        val mrot1  : Mat4 = Mat4.rotacionYgrad( -angulo_hor_grad )
        val mtras2 : Mat4 = Mat4.traslacion( Vec3( 0.0f, 0.0f, -distancia ))

        mat_vista = mtras2 * mrot2 * mrot1 * mtras1
    }
    // ---------------------------------------------------------------------------------------------

    /**
     * Modifica los ángulos horizonal y vertical de la cámara orbital
     *
     * @param incr_hor (number) -- incremento en horizontal
     * @param incr_ver (number) -- incremento en vertical
     */
    override fun mover( delta_x : Float, delta_y : Float )
    {
        val fac : Float = 0.3f
        this.angulo_hor_grad  += -fac*delta_x
        this.angulo_vert_grad += -fac*delta_y
    }
    // ---------------------------------------------------------------------------------------------

    /**
     * Modifica la distancia entre el punto de atención y el punto del observador
     * @param incr_z (number) +1 o -1 según la dirección de movimiento del ratón
     */
    override fun zoom( factor_rel : Float )
    {
        distancia = distancia/factor_rel
    }
}

// *************************************************************************************************


